### PR TITLE
Emoji Update

### PR DESCRIPTION
**Pull request by tmqCypher**
Finally got a chance to take a look at updating the emoji backend, apologies for the wait!

I made what seemed to be the most necessary changes talked about in #583:
- Removing unused information (names and descriptions) from the Emoji class
- Creating a Gradle task that generates a more efficient `res/raw/emojis.txt` file from the most recent Unicode standard
- Saving recently used emoji preferences as emoji values rather than names
- Migrating old user preferences to the new system

I kept the Emoji class because it keeps EmojiGridView and EmojiGroupButtonsBar cleaner and keeps space for potential features like #173, which would require more emoji-specific functionality than is offered by KeyValue.
<br>

The code does all of the above as it is now, but as part of the update I'd like to suggest saving recently used preferences as individual string-integer entries rather than a string set. We would no longer have to parse the integers out when the preferences are loaded or redundantly save unchanged values for every keystroke.

If this sounds like a good idea, or if any other preference changes should be made, they should definitely be implemented now to avoid a much more headache-inducing multi-stage migration process in the future.

Let me know what you think!

### CODE REVIEW

File: srcs/juloo.keyboard2/EmojiGridView.java [hunk 1]
```diff
@@ -110,6 +111,33 @@ SharedPreferences emojiSharedPreferences()
     return getContext().getSharedPreferences("emoji_last_use", Context.MODE_PRIVATE);
   }
 
+  private void migrateOldPrefs()
```
Comment (Julow) [inline srcs/juloo.keyboard2/EmojiGridView.java:114]: This does some work that is re-done quickly after and loose the result (`emojiSharedPreferences`, `getStringSet`).
What do you think of changing that into a function that returns the last used (eg. `Set<String> loadLastUsedFromPrefs()`) and would be called from `loadLastUsed()` ?
That would make this function the only place where the prefs are loaded, ensuring migration is always checked.

File: srcs/juloo.keyboard2/EmojiGridView.java [hunk 2]
```diff
@@ -110,6 +111,33 @@ SharedPreferences emojiSharedPreferences()
...
+    if (lastUsed != null && !prefs.getBoolean(MIGRATION_CHECK_KEY, false))
+    {
+      SharedPreferences.Editor edit = prefs.edit();
+      edit.clear();
+
+      Set<String> lastUsedNew = new HashSet<>();
+      for (String entry : lastUsed)
+      {
+        String[] data = entry.split("-", 2);
+        lastUsedNew.add(Integer.parseInt(data[0]) + "-" + Emoji.mapOldNameToValue(data[1]));
```
Comment (Julow) [inline srcs/juloo.keyboard2/EmojiGridView.java:132]: `mapOldNameToValue` can throw an exception and it's not fine to crash the app for such a small feature.
I think catching the exception here and not adding the problematic emoji to `lastUsedNew` is a nice way to handle it.

File: srcs/juloo.keyboard2/Emoji.java [hunk 3]
```diff
@@ -7,81 +7,782 @@
...
+      // Read emoji (until empty line)
+      while (!(line = reader.readLine()).isEmpty())
+      {
+        Emoji e = new Emoji(line);
+        _all.add(e);
+        _stringMap.put(line, e);
+      }
 
-  public static int num_groups = 0;
+      // Read group indices
```
Comment (Julow) [inline srcs/juloo.keyboard2/Emoji.java:50]: This format works but I think could be slightly simpler with the group indices on the first line. That way, the list don't have to be delimited.
The code that interprets the indices can still be done after the entire list has been read.

File: srcs/juloo.keyboard2/Emoji.java [hunk 4]
```diff
@@ -7,81 +7,782 @@
...
+  public static int getNumGroups()
+  {
+    return _groups.size();
+  }
 
-  public static Emoji getEmojiByName(String name)
+  public static Emoji[] getEmojisByGroup(int groupIndex)
   {
-    return emojis_by_name.get(name);
+    return _groups.get(groupIndex).toArray(new Emoji[0]);
```
Comment (Julow) [inline srcs/juloo.keyboard2/Emoji.java:68]: The List representation seems ideal for saving memory avoiding copies. Especially that this function is called for every group just to get the first emoji of each groups to make the tabs.
I feel like the rest of the code can be made to work with List with small modification, what do you think ?

**Review by Julow — COMMENTED** [review]
This is fantastic ! The code is very well done.

The APK do not decrease in size (and in fact increases by 85K), which is probably due to the to-be-removed migration function (the huge switch is 121K).

> The code does all of the above as it is now, but as part of the update I'd like to suggest saving recently used preferences as individual string-integer entries rather than a string set. We would no longer have to parse the integers out when the preferences are loaded or redundantly save unchanged values for every keystroke.

You mean having one StringSet representing the emojis that were used previously, and an extra integer preference for each of them ?
I think that could be a good idea !

`LayoutsPreference` also deals with serializing complex values and it uses [JSONObject](https://developer.android.com/reference/org/json/JSONObject) in order to rely less on `SharedPreference`'s methods.

There's also [getAll()](https://developer.android.com/reference/android/content/SharedPreferences#getAll()) but it's very unsafe and worrying about `LAST_USE_PREF` and `MIGRATION_CHECK_KEY` seems like a step backward.

File: srcs/juloo.keyboard2/Emoji.java (hunk 4 above)
Comment (tmqCypher) [reply srcs/juloo.keyboard2/Emoji.java:68]: Agreed. For the initial draft I wanted to make as few changes outside of Emoji.java as possible, but I can definitely refactor the rest to work with Lists

**Comment by tmqCypher** [comment]
> You mean having one StringSet representing the emojis that were used previously, and an extra integer preference for each of them ?

> There's also [getAll()](https://developer.android.com/reference/android/content/SharedPreferences#getAll()) but it's very unsafe and worrying about `LAST_USE_PREF` and `MIGRATION_CHECK_KEY` seems like a step backward.

My idea was to store only the integer preferences, use `getAll()` to retrieve the emoji-integer pairs, and copy the values to `_lastUsed`. This wouldn't require any de-serialization and would also remove the need for `MIGRATION_CHECK_KEY` since migration status could be determined simply by checking for the old string set with `prefs.contains(LAST_USE_REF)`.

All of the preferences would have a single, known type and nothing would be getting modified, so it doesn't seem unsafe to use `getAll()` in this case. I'm new to Android development though, so I'll trust your judgement if you'd still rather avoid it.

File: srcs/juloo.keyboard2/EmojiGridView.java (hunk 2 above)
Comment (tmqCypher) [reply srcs/juloo.keyboard2/EmojiGridView.java:131]: Whoops, thanks for noticing that. I made it throw for that exact purpose and forgot to actually catch it

**Comment by Julow** [comment]
Sorry for the slow reply. This should not be a draft anymore, I would like to merge it!

> My idea was to store only the integer preferences, use getAll() to retrieve the emoji-integer pairs, and copy the values to _lastUsed.

That works for me :) Though, the PR is working, so I suggest to merge it like this now and make further improvements later.

**Review by Julow — APPROVED** [review]
This will make our lives easier! Thanks a lot :)

**Comment by Julow** [comment]
The new preference layout you propose seems good to me but will require yet an other migration function as I want to avoid leaving this PR open for too long. I plan to release soon and need the updated emojis.

**Pull request merged**